### PR TITLE
#162685669 User can Favorite an Article

### PR DESCRIPTION
**Pull request by moseskamira**
#### What Does This PR Do?
This PR enables a user to favorite an article

#### Link To Pivotal Tracker Story
#162685669

#### How Should This Be Manually Tested
Fetching the branch 'ft-favorite-article-162685669' onto your local machine

#### Run the App

- Fetch the Article of interest by using the URL: http://127.0.0.1:8000/api/articles/slug/

- Using the PUT method, append the word 'favorite' on to the above URL to Favorite the Article: URL should appear as below; http://127.0.0.1:8000/api/articles/slug/favorite/

- To Retrieve All Favorited Articles, use the URL: http://127.0.0.1:8000/api/favorites/

#### NOTE
You do not need to pass anything in the body

##### Snapshots of expected results for both favoriting and unfavoriting an Article:

screen shot 2019-01-24 at 12 42 08 pm

screen shot 2019-01-25 at 7 41 31 am

### CODE REVIEW

**Review by tomuhenry — APPROVED** [review]
LGTM, I think the PR is running the way it is required to.

**Review by Opio-Emmanuel-Omona — APPROVED** [review]
LGTM

**Review by VivianDoreen — CHANGES_REQUESTED** [review]
Could you kindly use descriptive message when favouriting an article? Instead of returning the article, you could return "You have favourited the article"

image

**Comment by moseskamira** [comment]
> Could you kindly use descriptive message when favouriting an article? Instead of returning the article, you could return "You have favourited the article"
> 
> image

Worked upon your requested changes, could you please review and do the needful?
screen shot 2019-01-24 at 12 42 08 pm

**Review by VivianDoreen — APPROVED** [review]

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -97,3 +98,46 @@ def destroy(self, request, slug):
...
             status=status.HTTP_200_OK
         )
+
+
+class FavouritesView(GenericAPIView):
+    serializer_class = FavouriteSerializer
+    permission_classes = (IsAuthenticated,)
+    renderer_classes = (ArticleJSONRenderer,)
+
+    def put(self, request, **kwargs):
```
Comment (archbishop12) [inline authors/apps/articles/views.py:311]: @moseskamira Implement the feedback given to you during the review session.

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -97,3 +98,50 @@ def destroy(self, request, slug):
...
+        article = self.get_object()
+        user = request.user
+        article_id = article.id
+        fav = Favourites.objects.filter(
+            article_id=article_id, user=user
+        )
+        fav_option = fav.first()
+        fav.delete()
+        counter = article.favouriteCount
+        if (counter == 1):
```
Comment (archbishop12) [inline authors/apps/articles/views.py]: @moseskamira Is this if statement necessary. You can refactor the logic here.

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -97,3 +98,50 @@ def destroy(self, request, slug):
...
+            article.favourited = False
+        if (counter >= 1):
+            article.favouriteCount = counter - 1
+            article.save()
+            return Response({
+                "message":
+                    "article has been unfavorited"
+            }, status=200)
+        else:
+            serializer = self.serializer_class(data={}, partial=True)
```
Comment (archbishop12) [inline authors/apps/articles/views.py:339]: @moseskamira Look at the logic and make sure it is consistent. Ensure you don't `repeat yourself`.

**Comment by archbishop12** [comment]
@moseskamira When one tries to favourite an article they have to do it twice. Fix this flaw in your code.

**Pull request merged**